### PR TITLE
Make sampler length equal to the number of returned batches (#212)

### PR DESCRIPTION
**Pull request by janvainer**
Solves issue #212 

* Add _len attribute and _set_sampler_len method
* Add tests that check len(sampler) == len([batch for batch in sampler])

### CODE REVIEW

File: lhotse/dataset/sampling.py [hunk 1]
```diff
@@ -96,6 +98,9 @@ def _maybe_init_distributed(self, world_size: Optional[int], rank: Optional[int]
                 # A different error - pass it on.
                 raise
 
+    def _set_sampler_len(self):
+        self._len = sum(1 for item in self)
```
Comment (pzelasko) [inline lhotse/dataset/sampling.py:102]: Unfortunately, I don't think this works as expected. If you look at `__iter__`, it is going to shuffle the cuts every time it's called, so after this loop, the order of the cuts will be different.

Also this is not unset anywhere, so it is going to be incorrect in subsequent epochs.

File: test/dataset/test_sampling.py [hunk 2]
```diff
@@ -40,6 +40,11 @@ def test_single_cut_sampler_shuffling():
     # Invariant 3: the items are shuffled, i.e. the order is different than that in the CutSet
     assert sampler_cut_ids != [c.id for c in cut_set]
 
+    # check that len(sampler) gives consistent result
```
Comment (pzelasko) [inline test/dataset/test_sampling.py:43]: I think this test's inputs are insufficient to test this, as they all have equal lengths

Comment (janvainer) [reply test/dataset/test_sampling.py:43]: But that should be the point no? The lengths should equal

File: lhotse/dataset/sampling.py (hunk 1 above)
Comment (janvainer) [reply lhotse/dataset/sampling.py:102]: Hm so the shuffling should not happen in-place I guess?

File: test/dataset/test_sampling.py (hunk 2 above)
Comment (pzelasko) [reply test/dataset/test_sampling.py]: sorry I didn't express myself well... I meant that the cut inputs have equal lengths, so whatever is their order inside the sampler, you will always get the same len(sampler) with your changes, regardless of the epoch. If we create cut inputs with varying duration, then you can actually test whether len(sampler) is consistent in the same epoch and different in different epochs

File: lhotse/dataset/sampling.py (hunk 1 above)
Comment (janvainer) [reply lhotse/dataset/sampling.py:105]: @pzelasko is it necessary that the expression `list(sampler)` gives correct result in the case of samplers? The `list()` constructor behaves different from eg `[item for item in sampler]`. The former seems to call `sampler.__len__` asynchronously many times while the latter simply calls `__iter__` and `__next__` in sequential order as is expected. 
Under the current state of code, I am getting `list(sampler) == []` as can be seen from the test. Any idea how this could be surpassed?

Comment (pzelasko) [reply lhotse/dataset/sampling.py:105]: one option is to shuffle in a deterministic way, so that the same seed + epoch will always result in the same shuffling order (this actually makes sense and makes it possible to re-start training from a checkpoint mid-epoch in an exact way, as we can know how many batches to skip and from which batch to resume)

File: test/dataset/test_sampling.py (hunk 2 above)
Comment (janvainer) [reply test/dataset/test_sampling.py]: Yup I figured in the end :) The updated version uses varying lengths.

File: lhotse/dataset/sampling.py (hunk 1 above)
Comment (pzelasko) [reply lhotse/dataset/sampling.py:105]: Oh, I did not know that `list(x)` behaves this way. Anyway with fully deterministic shuffling that shouldn't be an issue anymore I think.

Comment (janvainer) [reply lhotse/dataset/sampling.py:105]: The shuffling you describe is basically how things work now I think. The iter method uses the self.seed + epoch random seed for shuffling so the process should be deterministic. I tried to determine the length already in init, but still getting this weird behavior.

Comment (pzelasko) [reply lhotse/dataset/sampling.py:105]: I don't think that's true, because we re-shuffle the list multiple times, so the elements are permuted in the same manner (because of the seed + epoch), but their order does change (just deterministically).

Comment (janvainer) [reply lhotse/dataset/sampling.py:105]: Sorry I dont follow. Maybe it i would understand if i see some code. How the current `__iter__` should change? Right now it looks like this
```
    def __iter__(self) -> 'CutSampler':
        """
        Prepare the dataset for iterating over a new epoch. Will shuffle the data if requested.
        """
        if self.shuffle:
            r = random.Random(self.seed + self.epoch)
            r.shuffle(self.data_source)
        self.current_idx = 0
        self._len = None
        return self

```

Comment (pzelasko) [reply lhotse/dataset/sampling.py:105]: OK take a look at the following example:
```python
>>> x = list(range(10))
>>> x
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
>>> r = Random(42)
>>> r.shuffle(x)
>>> x
[7, 3, 2, 8, 5, 6, 9, 4, 0, 1]
>>> r = Random(42)
>>> r.shuffle(x)
>>> x
[4, 8, 2, 0, 6, 9, 1, 5, 7, 3]
```

If you call iter multiple time, the same happens with cut ids (even though the seed + epoch are constant). We might need to hold the copy of the original cut ID ordering and re-compute the shuffling each time iter is called using the original order.

Comment (janvainer) [reply lhotse/dataset/sampling.py]: Ok I updated the code to remember the original ordering. A list of indices is remembered and shuffled instead of the cut_ids. The random permutation does not change unless the user specifies a different seed. Also the length gets calculated correctly (number of sampled batches == len(sampler)) and the `list()` constructor works too. Is this what you had on mind?

Comment (pzelasko) [reply lhotse/dataset/sampling.py]: Great! Yes that sounds right. I will review later or tomorrow. Thanks!

File: lhotse/dataset/sampling.py [hunk 3]
```diff
@@ -67,8 +84,13 @@ def __init__(
...
         self.seed = seed
         self.epoch = 0
+
         self._maybe_init_distributed(world_size=world_size, rank=rank)
-        self.data_source = partition_cut_ids(self.full_data_source, world_size=self.world_size, rank=self.rank)
+        self.data_source = DataSource(
+            partition_cut_ids(self.full_data_source, world_size=self.world_size, rank=self.rank)
+        )
+
+        self._len = None
```
Comment (pzelasko) [inline lhotse/dataset/sampling.py:93]: I suggest renaming this to `num_batches` everywhere, it seems more intuitive to set in the sub-classes (like `BucketingSampler`)

File: test/dataset/test_sampling.py [hunk 4]
```diff
@@ -136,6 +175,26 @@ def test_bucketing_sampler_cut_pairs():
     assert set(cut_set2.ids) == set(cut_ids)
 
 
+def test_bucketing_sampler_len():
+    # total duration is 55 seconds
```
Comment (pzelasko) [inline test/dataset/test_sampling.py:179]: not true for this test

File: test/dataset/test_sampling.py [hunk 5]
```diff
@@ -41,6 +41,25 @@ def test_single_cut_sampler_shuffling():
     assert sampler_cut_ids != [c.id for c in cut_set]
 
 
+def test_single_cut_sampler_len():
```
Comment (pzelasko) [inline test/dataset/test_sampling.py:44]: can these tests check that the batch ordering is different for every epoch too? you can just store the "batches" as lists and assert that they are not equal (should work since these are just collections of string ids, not dicts with actual data)

Comment (pzelasko) [reply test/dataset/test_sampling.py:44]: (this is mostly to avoid regressions in the future, I can see that the current code should work as expected)

Comment (pzelasko) [reply test/dataset/test_sampling.py:44]: I think it would be good to also have a test that shows that the order is deterministic given the same epoch number. This is not an obvious property in general and I wouldn't be surprised if we accidentally break it trying to update this code in the future.

**Review by pzelasko — COMMENTED** [review]
Thanks, it's great! Just one cosmetic comment and some notes about testing.

File: test/dataset/test_sampling.py (hunk 4 above)
Comment (janvainer) [reply test/dataset/test_sampling.py:179]: good catch, forgot to adjust

File: lhotse/dataset/sampling.py (hunk 3 above)
Comment (janvainer) [reply lhotse/dataset/sampling.py:93]: renaming

File: test/dataset/test_sampling.py (hunk 5 above)
Comment (janvainer) [reply test/dataset/test_sampling.py:79]: I added several tests for ordering and determinism. The sampling of buckets in bucket sampler is now also deterministic and dependent on seed and epoch. I also added seed parameter for the bucket sampler to be able to control the bucket sampling itself.

One more idea: We could add a parameter that would automatically increase epoch after iteration so that user does not have to manually call `set_epoch`. WYT?

Comment (janvainer) [reply test/dataset/test_sampling.py:79]: Hm and one more problem - the `list(sampler)` works fine in `python3.8` but not in `python3.7` for some reason. I don't understand what exactly is wrong.

Comment (pzelasko) [reply test/dataset/test_sampling.py:79]: > One more idea: We could add a parameter that would automatically increase epoch after iteration so that user does not have to manually call `set_epoch`. WYT?

I'd rather not - that would be inconsistent with what PyTorch's DistributedSampler does.

> Hm and one more problem - the list(sampler) works fine in python3.8 but not in python3.7 for some reason. I don't understand what exactly is wrong.

I wouldn't worry about that - if this is problematic later we can look into it.

**Comment by pzelasko** [comment]
Be careful with force-pushing - I merged master to check if it's merge-able

**Review by pzelasko — APPROVED** [review]

**Comment by pzelasko** [comment]
Thank you, great work!

File: lhotse/dataset/sampling.py [hunk 6]
```diff
@@ -367,6 +380,7 @@ def __iter__(self) -> 'BucketingSampler':
         return self
 
     def __next__(self) -> List[str]:
+        random.seed(self.seed + self.epoch)
```
Comment (pzelasko) [inline lhotse/dataset/sampling.py:383]: I missed this during the review, but it silently breaks the bucketing sampler - it will always choose the same bucket until depletion
